File: AOC_2022/Day 9/9.py
file = open("input.txt", "r")

lines = file.read().splitlines()

import numpy as np

head = (0, 0)
tail = (0, 0)
visited = set()
visited.add(tail)

for line in lines:
    direction, steps = line.split()
    for _ in range(int(steps)):
        if direction == 'U':
                head = (head[0], head[1] + 1)
        elif direction == 'D':
                head = (head[0], head[1] - 1)
        elif direction == 'R':
                head = (head[0] + 1, head[1])
        elif direction == 'L':
                head = (head[0] - 1, head[1])
        diff_x = head[0] - tail[0]
        diff_y = head[1] - tail[1]
        if abs(diff_x) > 1 or abs(diff_y) > 1:
            tail = (tail[0] + np.sign(diff_x), tail[1] + np.sign(diff_y))
            visited.add(tail)

print("The Result for the 1st star is: " + str(len(visited)))

rope = [(0, 0) for _ in range(10)]

visited = set()
visited.add((0, 0))

for line in lines:
    direction, steps = line.split()
    for _ in range(int(steps)):
        if direction == 'U':
            rope[0] = (rope[0][0], rope[0][1] + 1)
        elif direction == 'D':
            rope[0] = (rope[0][0], rope[0][1] - 1)
        elif direction == 'R':
            rope[0] = (rope[0][0] + 1, rope[0][1])
        elif direction == 'L':
            rope[0] = (rope[0][0] - 1, rope[0][1])
        for i in range(1,10):
            diff_x = rope[i][0] - rope[i-1][0]
            diff_y = rope[i][1] - rope[i-1][1]
            if abs(diff_x) > 1 or abs(diff_y) > 1:
                rope[i] = (rope[i][0] - np.sign(diff_x), rope[i][1] - np.sign(diff_y))
            
        visited.add(rope[9])

print("The Result for the 2nd star is: " + str(len(visited)))

File: AOC_2022/Day 9/9.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <cctype>
#include <vector>
#include <set>
#include <algorithm>
#include <cstdlib>
using namespace std;

#define sgn(v) ( ( (v) < 0 ) ? -1 : ( (v) > 0 ) )

int main()
{
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    string line;
    set<pair<int,int>> visited;
    pair<int,int> head = {0,0};
    pair<int,int> tail = {0,0};
    visited.insert(tail);
    while (getline(in, line))
    {
        char dir = line[0];
        int dist = stoi(line.substr(1));
        while(dist--){
            switch(dir){
                case 'U': head.second++; break;
                case 'D': head.second--; break;
                case 'L': head.first--; break;
                case 'R': head.first++; break;
            }
            int diff_x = head.first - tail.first,diff_y = head.second - tail.second;
            if(abs(diff_x) > 1 || abs(diff_y) > 1){
                tail.first += sgn(diff_x);
                tail.second += sgn(diff_y);
                visited.insert(tail);
            }
        }
    }
    cout << visited.size() << "\n";
    
    // Part 2
    in.clear();
    in.seekg(0, ios::beg);
    visited.clear();
    vector<pair<int,int>> rope(10, {0,0});
    visited.insert({0,0});
    while (getline(in, line))
    {
        char dir = line[0];
        int dist = stoi(line.substr(1));
        while(dist--){
            switch(dir){
                case 'U': rope[0].second++; break;
                case 'D': rope[0].second--; break;
                case 'L': rope[0].first--; break;
                case 'R': rope[0].first++; break;
            }
            for(int i = 1; i < 10;i++){
                int diff_x = rope[i].first - rope[i-1].first,diff_y = rope[i].second - rope[i-1].second;
                if(abs(diff_x) > 1 || abs(diff_y) > 1){
                    rope[i].first -= sgn(diff_x);
                    rope[i].second -= sgn(diff_y);    
                }
            }
            visited.insert(rope[9]);
        }
    }
    cout<<visited.size()<<"\n";
    return 0;
}
    